File: agent/library/src/main/java/org/matilda/services/plugins/PluginsService.java
package org.matilda.services.plugins;

import org.matilda.commands.*;

import javax.inject.Inject;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

@MatildaService
public class PluginsService {
    @Inject
    JavaLoader mJavaLoader;

    @Inject
    CommandRepository mCommandRepository;

    @Inject
    PluginDependencies mPluginDependencies;

    @Inject
    PluginsService() {}

    @MatildaCommand
    public int loadPlugin(byte[] jarBytes, String className) throws IOException, ClassNotFoundException,
            InvocationTargetException, IllegalAccessException, NoSuchMethodException {
        ClassLoader classLoader = mJavaLoader.load(jarBytes);
        Class<?> entryPointClass = Class.forName(className, true, classLoader);
        Method commandRegistryMethod = entryPointClass.getDeclaredMethod("createCommandRegistry",
                PluginDependenciesModule.class);
        CommandRegistry commandRegistry =
                (CommandRegistry) commandRegistryMethod.invoke(null, new PluginDependenciesModule(mPluginDependencies));
        return mCommandRepository.addCommandRegistry(commandRegistry);
    }
}

File: agent/library/src/main/java/org/matilda/services/plugins/PluginsService.kt
package org.matilda.services.plugins

import me.tatarka.inject.annotations.Inject
import org.matilda.commands.*

@MatildaService
class PluginsService @Inject constructor(private val mJavaLoader: JavaLoader,
                                         private val mCommandRepository: CommandRepository,
                                         private val mPluginDependencies: PluginDependencies) {
    @MatildaCommand
    fun loadPlugin(jarBytes: ByteArray, className: String): Int {
        val classLoader = mJavaLoader.load(jarBytes)
        val entryPointClass = Class.forName(className, true, classLoader)
        val commandRegistryMethod = entryPointClass.getDeclaredMethod(
            "createCommandRegistry",
            PluginDependencies::class.java
        )
        val commandRegistry =
            commandRegistryMethod.invoke(null, mPluginDependencies) as CommandRegistry
        return mCommandRepository.addCommandRegistry(commandRegistry)
    }
}
